elastic-agent-client.proto

```diff
@@ -112,15 +114,301 @@ message ActionResponse {
+}
+
+// Response from `DiscardTxn`.
+message StoreDiscardTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentStore {
+    // Key-Value state storage is provided for each unit.
+    //
+    // Transactional store is provided to allow multiple key operations to occur before a commit to ensure consistent
+    // state when multiple keys make up the state of an units persistent state.
+    rpc BeginTxn(StoreBeginTxnRequest) returns (StoreBeginTxnResponse);
+    rpc GetKey(StoreGetKeyRequest) returns (StoreGetKeyResponse);
+    rpc SetKey(StoreSetKeyRequest) returns (StoreSetKeyResponse);
+    rpc DeleteKey(StoreDeleteKeyRequest) returns (StoreDeleteKeyResponse);
+    rpc CommitTxn(StoreCommitRequest) returns (StoreCommitTxnResponse);
+    rpc DiscardTxn(StoreDiscardRequest) returns (StoreDiscardTxnResponse);
+}
+
+// Requests an artifact from the Elastic Agent.
+message ArtifactFetchRequest {
+    // Token that is used to uniquely identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+    // ID of the artifact.
+    string id = 2;
+    // SHA256 of the artifact.
+    string sha256 = 3;
+}
+
+// Content of the artifact.
+message ArtifactFetchResponse {
+    // Artifact content.
+    bytes content = 1;
+}
+
+service ElasticAgentArtifact {
+    // Fetches an artifact from the artifact store.
+    //
+    // Response from this call can be chunked over multiple `ArtifactFetchResponse` for very large responses. A minimum
+    // of two responses will always be returned. The last response is always empty to signal that its the end of the
+    // response.
+    rpc Fetch(ArtifactFetchRequest) returns (stream ArtifactFetchResponse);
+}
+
+
+// A single metric that is reported by a unit. A unit can report multiple metrics in the same period.
+message UnitMetric {
+    enum MetricType {
+        KEYWORD = 0;
+        LONG = 1;
+    }
+
+    // Label for the metric.
+    string label = 1;
+
+    // Type for the metric.
+    MetricType type = 2;
+
+    // Value for the metric.
+    bytes value = 3;
+}
+
+message UnitMetrics {
+    // ID of the unit.
+    string id = 1;
+
+    // Metrics for this unit in this period.
+    repeated UnitMetric metrics = 2;
+}
+
+message MetricsReportRequest {
+    // Token that is used to uniquely identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+
+    // Period of time for the metrics that are collected.
+    uint64 period = 2;
+
+    // Reported units for this period.
+    repeated UnitMetrics units = 3;
+}
+
+// Response from `MetricsReport`.
+message MetricsReportResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentMetrics {
+    // Report metrics to the Elastic Agent.
+    rpc Report(MetricsReportRequest) returns (MetricsReportResponse);
```

Comment:
Where do these metrics go? I am curious how you avoid re-implementing parts of metricbeat or the data plane in general in the long term.

There was some consideration given to adding a generalized metric endpoint to the shipper, supporting an OpenTelemetry metrics or Statsd endpoint or similar there would have value for the future metricbeat+integrations beyond just monitoring agent processes.

Reply:
@cmacknz @blakerouse Let's have a brainstorm on this to to make sure we are aligned on the goal.  I am not too worried about reimplementing the data plane, I would consider these metrics, agent health information and not something that metricbeat would collect.